### ISSUE TITLE
Remove properties that are used by invalid code

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#4759]
Currently we have the following AST properties declared:

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSEnumDeclaration/spec.ts#L13

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSModuleDeclaration/spec.ts#L25

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSInterfaceDeclaration/spec.ts#L14

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSInterfaceDeclaration/spec.ts#L15

These all represent bad code that would cause errors if one attempted to build in TS:
```
public enum Foo {}
// 'public' modifier cannot appear on a module or namespace element.(1044)

abstract namespace Bar {}
// 'abstract' modifier can only appear on a class, method, or property declaration.(1242)

interface Baz implements Bam {}
// Interface declaration cannot have 'implements' clause.(1176)

abstract interface Bam {}
// 'abstract' modifier can only appear on a class, method, or property declaration.
```

TS parses these fine and emits a semantic error, not a syntactic error.
Our parser currently does not error on syntactically bad code, thus we also don't error on it (#1852).

We should remove these properties from the AST as it is non-sensical to include them - nobody should ever be inspecting the code that is produced on them.

**Comment by Josh-Cena** [comment]
Remove properties that...?

**Comment by DMartens** [comment]
I intented to create an PR for this, but each of the properties seem to be intended as they are handled explicitly:
- [TSEnumDeclaration and TSModuleDeclaration](https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/typescript-estree/src/convert.ts#L681) have an own private function in the converter to the modifiers
- [TSInterfaceDeclaration](https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/typescript-estree/src/convert.ts#L2641) explicitly sets `implements` if it is available

**Comment by juank1809** [comment]
I think it is not correct to remove these properties, as there is still valid properties (or at least one)  for enum and a module declaration:

```javascript
declare interface enum Foo {} 
```

**Comment by juank1809** [comment]
I was about to submit a PR to this issue, but I stumbled upon a realization:  there are still [valid modifiers](https://github.com/microsoft/TypeScript/blob/main/lib/typescript.d.ts#L621), more specifically _const_ and _declare_, to handle

I think the best approach to this problem is to modify the code in order to restrict and forbid invalid modifiers within these nodes.

**Comment by bradzacher** [comment]
> more specifically const and declare

These are handled via boolean flags.

https://github.com/typescript-eslint/typescript-eslint/blob/18a81cb715d994e8527607c8ee3fe848a59922d8/packages/ast-spec/src/declaration/TSEnumDeclaration/spec.ts#L11-L12

https://github.com/typescript-eslint/typescript-eslint/blob/18a81cb715d994e8527607c8ee3fe848a59922d8/packages/ast-spec/src/declaration/TSModuleDeclaration/spec.ts#L24

https://github.com/typescript-eslint/typescript-eslint/blob/18a81cb715d994e8527607c8ee3fe848a59922d8/packages/ast-spec/src/declaration/TSInterfaceDeclaration/spec.ts#L16

**Comment by bradzacher** [comment]
Another set of cases to be removed:
https://github.com/typescript-eslint/typescript-eslint/blob/8ed72192c274249d26628fb125796e71318b857a/packages/ast-spec/src/element/TSIndexSignature/spec.ts#L13
https://github.com/typescript-eslint/typescript-eslint/blob/8ed72192c274249d26628fb125796e71318b857a/packages/ast-spec/src/element/TSMethodSignature/spec.ts#L23
https://github.com/typescript-eslint/typescript-eslint/blob/8ed72192c274249d26628fb125796e71318b857a/packages/ast-spec/src/element/TSPropertySignature/spec.ts#L21
https://github.com/typescript-eslint/typescript-eslint/blob/8ed72192c274249d26628fb125796e71318b857a/packages/ast-spec/src/parameter/TSParameterProperty/spec.ts#L14

Syntactically the TS parser allows the `export` modifier, but semantically it disallows it.

**Comment by bradzacher** [comment]
Another case to be removed:

https://github.com/typescript-eslint/typescript-eslint/blob/8ed72192c274249d26628fb125796e71318b857a/packages/ast-spec/src/element/TSPropertySignature/spec.ts#L18

Syntactically the TS parser allows an initialiser (likely due to it reusing parsing logic for class properties), but semantically it is disallowed.

**Comment by bradzacher** [comment]
Another case to be removed:
https://github.com/typescript-eslint/typescript-eslint/blob/8ed72192c274249d26628fb125796e71318b857a/packages/ast-spec/src/base/MethodDefinitionBase.ts#L24

`typeParameters` is placed on the `.value` node, not the `MethodDefinition`

**Comment by JoshuaKGoldberg** [comment]
Fixed by #4863 and will be available in v6! 🚀
